src/basics.md

```diff
@@ -122,4 +122,8 @@ function doSomething(foo: Foo): void {
 }
 ```
 
+### Non-linear compilation
+
+AssemblyScript does not compile a module linearly, but starts at the module's exports and only compiles what's reachable from them, often referred to as tree-shaking. As such, dead code is always validated syntactically, but not necessarily checked for semantic correctness. While this mechanism significantly helps to reduce compile times and feels almost natural to those familiar with *executing* JavaScript, it may initially feel a little strange not only to those with a background in traditional compilers, for example because emitted diagnostics are not linear, but also to those with a background in TypeScript, because even type annotations remain unchecked when unused. The exception to the rule is top-level code, including top-level variable declarations and their initializers, that must be evaluated in proper order.
```

Comment:
How about mention similar approaches like [non-strict semantics in haskell](https://wiki.haskell.org/Non-strict_semantics)?

Reply:
I am not 100% sure how much of this is non-strict semantics actually, given the examples shown there. Omitted mentioning it because I don't know enough to be confident about it.

Reply:
Yeah, reread this and it seems more relate to evaluation rather than parsing / compiling